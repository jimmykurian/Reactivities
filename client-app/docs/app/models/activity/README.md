[**Client App Documentation v0.0.0**](../../../README.md) â€¢ **Docs**

***

[Client App Documentation v0.0.0](../../../README.md) / app/models/activity

# app/models/activity

## Index

### Interfaces

- [Activity](interfaces/Activity.md)
