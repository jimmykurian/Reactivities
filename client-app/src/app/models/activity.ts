/**
 * @author Jimmy Kurian
 * @name Activity
 */

/**
 * The Activity interface represents the structure of an activity object in the Reactivities application.
 * It includes various properties such as id, title, date, description, category, city, and venue.
 *
 * @interface
 *
 * @property {string} id - The unique identifier for the activity.
 * @property {string} title - The title of the activity.
 * @property {string} date - The date of the activity in ISO string format.
 * @property {string} description - A brief description of the activity.
 * @property {string} category - The category to which the activity belongs.
 * @property {string} city - The city where the activity takes place.
 * @property {string} venue - The specific venue of the activity within the city.
 */
export interface Activity {
  id: string;
  title: string;
  date: string;
  description: string;
  category: string;
  city: string;
  venue: string;
}
