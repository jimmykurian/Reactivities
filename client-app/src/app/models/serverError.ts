/**
 * @file ServerError.ts
 * @author Jimmy Kurian
 * @fileoverview This file defines the `ServerError` interface, which models the structure of server error responses in the Reactivities application.
 */

/**
 * The ServerError interface represents the structure of an error returned from the server.
 * It includes the HTTP status code, an error message, and optional detailed information about the error.
 *
 * @interface ServerError
 *
 * @property {number} statusCode - The HTTP status code associated with the error.
 * @property {string} message - A brief message describing the error.
 * @property {string} details - Additional details or stack trace information about the error (optional).
 */
export interface ServerError {
  statusCode: number;
  message: string;
  details: string;
}
