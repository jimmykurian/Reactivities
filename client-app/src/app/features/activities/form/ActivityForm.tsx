/**
 * @file ActivityForm.tsx
 * @author Jimmy Kurian
 * @fileoverview This file contains the ActivityForm component.
 */

import { Button, Form, Segment } from 'semantic-ui-react';
import { useEffect, useState } from 'react';
import { useStore } from '../../../stores/store';
import { observer } from 'mobx-react-lite';
import { useNavigate, useParams } from 'react-router-dom';
import { Activity } from '../../../models/activity';
import LoadingComponent from '../../../layout/LoadingComponent';
import { v4 as uuid } from 'uuid';

/**
 * @component ActivityForm
 * @description A functional component that renders a form for activity details.
 *
 * @returns {JSX.Element} The JSX element representing the activity form.
 *
 * @remarks
 * This component uses the Semantic UI React library to create a form layout
 * for collecting activity details such as title, description, category, date, city, and venue.
 * It includes input fields and buttons for submitting or canceling the form.
 * The form state is managed using the `useState` hook, and the `handleInputChange` function updates the form state.
 * The `handleSubmit` function is used to create or edit the activity when the form is submitted.
 * The component accesses the `activityStore` from the MobX store context to get the selected activity, createActivity, updateActivity, and loading functions.
 * The `loadActivity` function is used to load the activity details when the component is mounted.
 * The `LoadingComponent` is displayed while the activity details are being loaded.
 *
 * @example
 * Here is an example of how to use the ActivityForm component:
 * ```tsx
 * import React from 'react';
 * import { BrowserRouter as Router, Route } from 'react-router-dom';
 * import { StoreProvider } from './stores/store';
 * import ActivityForm from './features/activities/form/ActivityForm';
 *
 * const App = () => (
 *   <StoreProvider>
 *     <Router>
 *       <Route path="/manage/:id?" component={ActivityForm} />
 *     </Router>
 *   </StoreProvider>
 * );
 *
 * export default App;
 * ```
 */
export default observer(function ActivityForm(): JSX.Element {
  const { activityStore } = useStore();
  const {
    createActivity,
    updateActivity,
    loading,
    loadActivity,
    loadingInitial,
  } = activityStore;
  const { id } = useParams();
  const navigate = useNavigate();

  const [activity, setActivity] = useState<Activity>({
    id: '',
    title: '',
    date: '',
    description: '',
    category: '',
    city: '',
    venue: '',
  });

  useEffect(() => {
    if (id) loadActivity(id).then((activity) => setActivity(activity!));
  }, [id, loadActivity]);

  /**
   * @function handleSubmit
   * @description Handles form submission. Calls the create or update function based on the activity ID.
   */
  function handleSubmit(): void {
    if (!activity.id) {
      activity.id = uuid();
      createActivity(activity).then(() =>
        navigate(`/activities/${activity.id}`),
      );
    } else {
      updateActivity(activity).then(() =>
        navigate(`/activities/${activity.id}`),
      );
    }
  }

  /**
   * @function handleInputChange
   * @description Handles input changes in the form fields and updates the state.
   * @param {React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>} event - The event triggered by changing the form input fields.
   */
  function handleInputChange(
    event: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>,
  ): void {
    const { name, value } = event.target;
    setActivity({ ...activity, [name]: value });
  }

  if (loadingInitial) return <LoadingComponent content="Loading activity..." />;

  return (
    <Segment clearing>
      <Form onSubmit={handleSubmit} autoComplete="off">
        <Form.Input
          placeholder="Title"
          value={activity.title}
          name="title"
          onChange={handleInputChange}
        />
        <Form.TextArea
          placeholder="Description"
          value={activity.description}
          name="description"
          onChange={handleInputChange}
        />
        <Form.Input
          placeholder="Category"
          value={activity.category}
          name="category"
          onChange={handleInputChange}
        />
        <Form.Input
          placeholder="Date"
          type="date"
          value={activity.date}
          name="date"
          onChange={handleInputChange}
        />
        <Form.Input
          placeholder="City"
          value={activity.city}
          name="city"
          onChange={handleInputChange}
        />
        <Form.Input
          placeholder="Venue"
          value={activity.venue}
          name="venue"
          onChange={handleInputChange}
        />
        <Button
          loading={loading}
          floated="right"
          positive
          type="submit"
          content="Submit"
        />
        <Button floated="right" type="button" content="Cancel" />
      </Form>
    </Segment>
  );
});
