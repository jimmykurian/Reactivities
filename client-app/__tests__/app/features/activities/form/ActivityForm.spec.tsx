import { render, screen, fireEvent } from '@testing-library/react';
import { faker } from '@faker-js/faker';
import ActivityForm from '../../../../../src/app/features/activities/form/ActivityForm';
import { Activity } from '../../../../../src/app/models/activity';

// Static mock activity for for snapshot test
const staticMockActivity: Activity = {
  id: '1',
  title: 'Static Title',
  date: '2024-06-12T00:00:00.000Z',
  description: 'This is a static description for testing.',
  category: 'static-category',
  city: 'Static City',
  venue: 'Static Venue',
};

// Dynamic mock activity for normal unit tests
const generateMockActivity = (): Activity => ({
  id: faker.string.uuid(),
  title: faker.lorem.words(3),
  date: faker.date.future().toISOString().split('T')[0],
  description: faker.lorem.sentence(),
  category: faker.lorem.word(),
  city: faker.location.city(),
  venue: faker.location.streetAddress(),
});

describe('ActivityForm', () => {
  const closeForm = jest.fn();
  const createOrEdit = jest.fn();

  test('renders the ActivityForm component', () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <ActivityForm
        activity={mockActivity}
        closeForm={closeForm}
        createOrEdit={createOrEdit}
      />,
    );

    // Act & Assert
    expect(screen.getByPlaceholderText('Title')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Description')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Category')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Date')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('City')).toBeInTheDocument();
    expect(screen.getByPlaceholderText('Venue')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Submit/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Cancel/i })).toBeInTheDocument();
  });

  test('calls submit handler when form is submitted', () => {
    // Arrange
    const mockActivity = generateMockActivity();
    render(
      <ActivityForm
        activity={mockActivity}
        closeForm={closeForm}
        createOrEdit={createOrEdit}
      />,
    );

    // Act
    fireEvent.click(screen.getByRole('button', { name: /Submit/i }));

    // Assert
    expect(createOrEdit).toHaveBeenCalledTimes(1);
  });

  test('matches snapshot', () => {
    // Arrange
    const { asFragment } = render(
      <ActivityForm
        activity={staticMockActivity}
        closeForm={closeForm}
        createOrEdit={createOrEdit}
      />,
    );

    // Act & Assert
    expect(asFragment()).toMatchSnapshot();
  });
});
