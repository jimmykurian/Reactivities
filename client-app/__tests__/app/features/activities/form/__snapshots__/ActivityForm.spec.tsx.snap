// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui clearing segment"
  >
    <form
      autocomplete="off"
      class="ui form"
    >
      <div
        class="field"
      >
        <div
          class="ui input"
        >
          <input
            name="title"
            placeholder="Title"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <textarea
          name="description"
          placeholder="Description"
          rows="3"
        />
      </div>
      <div
        class="field"
      >
        <div
          class="ui input"
        >
          <input
            name="category"
            placeholder="Category"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <div
          class="ui input"
        >
          <input
            name="date"
            placeholder="Date"
            type="date"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <div
          class="ui input"
        >
          <input
            name="city"
            placeholder="City"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="field"
      >
        <div
          class="ui input"
        >
          <input
            name="venue"
            placeholder="Venue"
            type="text"
            value=""
          />
        </div>
      </div>
      <button
        class="ui positive right floated button"
        type="submit"
      >
        Submit
      </button>
      <a
        class="ui right floated button"
        href="/activities"
        role="button"
        type="button"
      >
        Cancel
      </a>
    </form>
  </div>
</DocumentFragment>
`;
